README.md
formality
============


[![Project Status: Active - The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active)
[![Build
Status](https://travis-ci.org/trinker/formality.svg?branch=master)](https://travis-ci.org/trinker/formality)
[![Coverage
Status](https://coveralls.io/repos/trinker/formality/badge.svg?branch=master)](https://coveralls.io/r/trinker/formality?branch=master)
<a href="https://img.shields.io/badge/Version-0.1.0-orange.svg"><img src="https://img.shields.io/badge/Version-0.1.0-orange.svg" alt="Version"/></a>
</p>
<img src="inst/formality_logo/r_formality.png" width="300" alt="tagger Logo">

**formality** utilizes the
[**tagger**](https://github.com/trinker/tagger) package to conduct
formality analysis. Heylighen (1999) and Heylighen & Dewaele (2002,
1999) have given the *F-measure* as a measure of how *contextual* or
*formal* language is. Language is considered more formal when it
contains much of the information directly in the text, whereas,
contextual language relies on shared experiences to more efficiently
dialogue with others.


Table of Contents
============

-   [Formality Equation](#formality-equation)
-   [Installation](#installation)
-   [Contact](#contact)
-   [Examples](#examples)
    -   [Load the Tools/Data](#load-the-toolsdata)
    -   [Assessing Formality](#assessing-formality)
    -   [Recycling the First Run](#recycling-the-first-run)
    -   [Plotting](#plotting)

Formality Equation
============


The **formality** package's main function is also titled `formality` and
uses Heylighen & Dewaele's (1999) *F-measure*. The *F-measure* is
defined formally as:

*F* = 50(((*n*<sub>*f*</sub> − *n*<sub>*c*</sub>)/*N*) + 1)

Where:

*f* = {*n**o**u**n*, *a**d**j**e**c**t**i**v**e*, *p**r**e**p**o**s**i**t**i**o**n*, *a**r**t**i**c**l**e*}
  
*c* = {*p**r**o**n**o**u**n*, *v**e**r**b*, *a**d**v**e**r**b*, *i**n**t**e**r**j**e**c**t**i**o**n*}
  
*N* = *n*<sub>*f*</sub> + *n*<sub>*c*</sub>

This yields an *F-measure* between 0 and 100%, with completely
contextualized language on the zero end and completely formal language
on the 100 end.

Please see the following references for more details about formality and
the *F-measure*:

-   Heylighen, F. (1999). Advantages and limitations of formal
    expression. Foundations of Science, 4, 25-56.
    <a href="http://link.springer.com/article/10.1023%2FA%3A1009686703349">doi:10.1023/A:1009686703349</a>
-   Heylighen, F. & Dewaele, J.-M. (1999). Formality of language:
    Definition, measurement and behavioral determinants. Center "Leo
    Apostel", Free University of Brussels. Retrieved from
    [<http://pespmc1.vub.ac.be/Papers/Formality.pdf>](http://pespmc1.vub.ac.be/Papers/Formality.pdf)
-   Heylighen, F. & Dewaele, J.-M. (2002). Variation in the
    contextuality of language: An empirical measure. Foundations of
    Science, 7(3), 293-340.
    <a href="http://link.springer.com/article/10.1023%2FA%3A1019661126744">doi:10.1023/A:1019661126744</a>

Installation
============

To download the development version of **formality**:

Download the [zip
ball](https://github.com/trinker/formality/zipball/master) or [tar
ball](https://github.com/trinker/formality/tarball/master), decompress
and run `R CMD INSTALL` on it, or use the **pacman** package to install
the development version:

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load_gh(c(
        "trinker/termco", 
        "trinker/tagger", 
        "trinker/formality"
    ))

Contact
=======

You are welcome to: 
* submit suggestions and bug-reports at: <https://github.com/trinker/formality/issues> 
* send a pull request on: <https://github.com/trinker/formality/> 
* compose a friendly e-mail to: <tyler.rinker@gmail.com>


Examples
========

The following examples demonstrate some of the functionality of
**formality**.

Load the Tools/Data
-------------------

    library(formality)
    data(presidential_debates_2012)

Assessing Formality
-------------------

`formality` takes the text as `text.var` and any number of grouping
variables as `grouping.var`. Here we use the `presidential_debates_2012`
data set and look at the formality of the people involved. Note that for
smaller text Heylighen & Dewaele (2002) state:

> At present, a sample would probably need to contain a few hundred
> words for the measure to be minimally reliable. For single sentences,
> the F-value should only be computed for purposes of illustration" (p.
> 24).

    form1 <- with(presidential_debates_2012, formality(dialogue, person))
    form1

    ##       person noun preposition adjective article verb pronoun adverb
    ## 1:  QUESTION  155          91        70      38  112      77     26
    ## 2:    LEHRER  182         104        93      62  164     101     48
    ## 3: SCHIEFFER  347         209       176     102  342     211     69
    ## 4:    ROMNEY 4406        3178      2346    1396 4676    2490   1315
    ## 5:     OBAMA 3993        2909      1935    1070 4593    2418   1398
    ## 6:   CROWLEY  387         269       135     104  405     249    134
    ##    interjection formal contextual     n        F
    ## 1:            4    354        219   573 61.78010
    ## 2:            8    441        321   762 57.87402
    ## 3:            0    834        622  1456 57.28022
    ## 4:           25  11326       8506 19832 57.10972
    ## 5:           13   9907       8422 18329 54.05096
    ## 6:            0    895        788  1683 53.17885

Recycling the First Run
-----------------------

This will take ~20 seconds because of the part of speech tagging that
must be undertaken. The output can be reused as `text.var`, cutting the
time to a fraction of the first run.

    with(presidential_debates_2012, formality(form1, list(time, person)))

    ##       time    person noun preposition adjective article verb pronoun
    ##  1: time 2  QUESTION  155          91        70      38  112      77
    ##  2: time 1    LEHRER  182         104        93      62  164     101
    ##  3: time 1    ROMNEY  950         642       483     286  978     504
    ##  4: time 3    ROMNEY 1766        1388       958     617 1920    1029
    ##  5: time 3 SCHIEFFER  347         209       176     102  342     211
    ##  6: time 2    ROMNEY 1690        1148       905     493 1778     957
    ##  7: time 3     OBAMA 1546        1185       741     432 1799     973
    ##  8: time 1     OBAMA  792         579       357     219  925     452
    ##  9: time 2     OBAMA 1655        1145       837     419 1869     993
    ## 10: time 2   CROWLEY  387         269       135     104  405     249
    ##     adverb interjection formal contextual    n        F
    ##  1:     26            4    354        219  573 61.78010
    ##  2:     48            8    441        321  762 57.87402
    ##  3:    240            4   2361       1726 4087 57.76853
    ##  4:    536           10   4729       3495 8224 57.50243
    ##  5:     69            0    834        622 1456 57.28022
    ##  6:    539           11   4236       3285 7521 56.32230
    ##  7:    522            4   3904       3298 7202 54.20716
    ##  8:    281            2   1947       1660 3607 53.97838
    ##  9:    595            7   4056       3464 7520 53.93617
    ## 10:    134            0    895        788 1683 53.17885

Plotting
--------

The generic `plot` function provides three views of the data:

1.  A filled bar plot of formal vs. contextual usage
2.  A dotplot of formality\*\*
3.  A heatmap of the usage of the parts of speech used to calculate the
    formality score

\*\****Note*** *red dot in center is a warning of less than 300 words*

    plot(form1)

![](inst/figure/unnamed-chunk-6-1.png)

The `plot` function uses **gridExtra** to stitch the plots together,
which is plotted imediately. However, the three subplots are actually
returned as a list as seen below.

    names(plot(form1, plot=FALSE))

    ## [1] "formality"         "contextual_formal" "pos"

Each of these is a **ggplot2** object that can be further manipulated
with various scales, facets, and annotations. I demonstrate some of this
functionality in the plots below.

    library(ggplot2)
    plot(form1, plot=FALSE)[[1]] +
        scale_size(range= c(8, 45)) +
        scale_x_continuous(limits = c(52, 63))

![](inst/figure/unnamed-chunk-8-1.png)

    plot(form1, plot=FALSE)[[2]] +
        scale_fill_grey()

![](inst/figure/unnamed-chunk-8-2.png)

    plot(form1, plot=FALSE)[[2]] +
        scale_fill_brewer(palette = "Pastel1") +
        facet_grid(~type)

![](inst/figure/unnamed-chunk-8-3.png)

    plot(form1, plot=FALSE)[[3]] +
        scale_fill_gradient(high = "red", low="white") +
        ggtitle("Participant's Use of Parts of Speech")

![](inst/figure/unnamed-chunk-8-4.png)

    plot(form1, plot=FALSE)[[3]] +
        scale_fill_gradient2(midpoint=.12, high = "red", low="blue")

![](inst/figure/unnamed-chunk-8-5.png)

NEWS
NEWS 
====

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor 
  and patch)
* New additions without breaking backward compatibility bumps the minor 
  (and resets the patch)
* Bug fixes and misc changes bumps the patch


formality 0.0.1 
----------------------------------------------------------------

This package is...


tests/testthat.R
library("testthat")
library("formality")

test_check("formality")

tests/testthat/test-formality.R
context("Checking formality")

test_that("formality ...",{


})



R/formality-package.R
#' 2012 U.S. Presidential Debates
#'
#' A dataset containing a cleaned version of all three presidential debates for
#' the 2012 election.
#'
#' @details
#' \itemize{
#'   \item person. The speaker
#'   \item tot. Turn of talk
#'   \item dialogue. The words spoken
#'   \item time. Variable indicating which of the three debates the dialogue is from
#' }
#'
#' @docType data
#' @keywords datasets
#' @name presidential_debates_2012
#' @usage data(presidential_debates_2012)
#' @format A data frame with 2912 rows and 4 variables
NULL



#' Sam I Am Text
#'
#' A dataset containing a character vector of the text from Seuss's 'Sam I Am'.
#'
#' @docType data
#' @keywords datasets
#' @name sam_i_am
#' @usage data(sam_i_am)
#' @format A character vector with 169 elements
#' @references Seuss, Dr. (1960). Green Eggs and Ham.
NULL


R/formality.R
#' Formality Score
#'
#' Formality score by grouping variable(s).
#'
#' @param text.var The text string variable or another \code{formality} object.
#' The latter recycles the part of speech tagging and is much faster.
#' @param grouping.var The grouping variable(s).  Default \code{NULL} generates
#' one word list for all text.  Also takes a single grouping variable or a list
#' of 1 or more grouping variables.  If \code{TRUE} an \code{id} variable is
#' used with a \code{seq_along} the \code{text.var}.
#' @param order.by.formality logical.  If \code{TRUE} orders the results
#' descending by formality score.
#' @param \ldots ignored.
#' @section Warning: Heylighen & Dewaele (2002) state, "At present, a sample would
#' probably need to contain a few hundred words for the measure to be minimally
#' reliable. For single sentences, the F-value should only be computed for
#' purposes of illustration" (p. 24).
#' @details Heylighen & Dewaele(2002)'s formality \emph{F-measure} is defined
#' formally as:
#'
#' \deqn{F = 50(\frac{n_{f}-n_{c}}{N} + 1)}{F = 50(n_f-n_c/N + 1)}
#'
#' Where:
#'
#' \deqn{f = \left \{noun, \;adjective, \;preposition, \;article\right \}}{f = {noun,adjective, preposition, article}}
#' \deqn{c = \left \{pronoun, \;verb, \;adverb, \;interjection\right \}}{c = {pronoun, verb, adverb, interjection}}
#' \deqn{N = \sum{(n_{f}\;+ \;n_{c})}}{N = \sum(f + c)}
#'
#' This yields an \emph{F-measure} between \eqn{0} and \eqn{100}\%, with
#' completely contextualizes language on the zero end and completely formal
#' language on the \eqn{100} end.
#' @references Heylighen, F. (1999). Advantages and limitations of formal expression. doi:10.1023/A:1009686703349 \cr \cr
#' Heylighen, F. & Dewaele, J.-M. (1999). Formality of language: Definition, measurement and behavioral determinants. Center "Leo Apostel", Free University of Brussels. Retrieved from \url{http://pespmc1.vub.ac.be/Papers/Formality.pdf} \cr \cr
#' Heylighen, F. & Dewaele, J.-M. (2002). Variation in the contextuality of language: An empirical measure. \emph{Foundations of Science}, 7(3), 293-340. doi:10.1023/A:1019661126744
#' @keywords formality
#' @importFrom data.table :=
#' @export
#' @examples
#' data(presidential_debates_2012)
#' (form1 <- with(presidential_debates_2012, formality(dialogue, person)))
#' with(presidential_debates_2012, formality(form1, list(person, time))) #recycle form 1 for speed
#'
#' plot(form1)
#' plot(with(presidential_debates_2012, formality(form1, list(person, time))))
formality <- function(text.var, grouping.var = NULL, order.by.formality = TRUE, ...){

    UseMethod("formality")

}

#' @export
#' @method formality default
formality.default <- function(text.var, grouping.var = NULL, order.by.formality = TRUE, ...){

    .SD <- n <- F <- NULL

    if(is.null(grouping.var)) {
        G <- "all"
    } else {
        if (is.list(grouping.var)) {
            m <- unlist(as.character(substitute(grouping.var))[-1])
            G <- sapply(strsplit(m, "$", fixed=TRUE), function(x) {
                    x[length(x)]
                }
            )
        } else {
            if (isTRUE(grouping.var)) {
                G <- "id"
            } else {
                G <- as.character(substitute(grouping.var))
                G <- G[length(G)]
            }
        }
    }
    if(is.null(grouping.var)){
        grouping <- rep("all", length(text.var))
    } else {
        if (isTRUE(grouping.var)) {
                grouping <- seq_along(text.var)
        } else {
            if (is.list(grouping.var) & length(grouping.var)>1) {
                grouping <- grouping.var
            } else {
                grouping <- unlist(grouping.var)
            }
        }
    }

    formal <- c('noun', 'preposition', 'adjective', 'article')
    contextual <- c('verb', 'pronoun', 'adverb', 'interjection')

    ## in other version this will be extracted
    #=============================================
    counts <- tagger::count_tags(tagger::as_basic(tagger::tag_pos(text.var)))
    data.table::setDT(counts)
    forms <- formal[formal %in% colnames(counts)]
    contexts <- contextual[contextual %in% colnames(counts)]
    cols <- c(forms, contexts)

    ## quoting for use in the data.table
    forms_exp <- parse(text=paste(forms, collapse=" + "))[[1]]
    contexts_exp <- parse(text=paste(contexts, collapse=" + "))[[1]]

    counts <- counts[, cols, with = FALSE][, formal := eval(forms_exp)][, contextual := eval(contexts_exp)]

    #=============================================
    ## in other version this will be extracted

    formality_counts <- new.env(hash=FALSE)
    formality_counts[["counts"]] <- counts

    counts <- cbind(
        data.table::as.data.table(stats::setNames(data.frame(grouping, stringsAsFactors = FALSE), G)),
        counts
    )

    out <- counts[, lapply(.SD, sum, na.rm = TRUE), keyby = G][,
        n := formal + contextual][,
        F := 50*(((formal - contextual)/(n))+1)]

    if (isTRUE(order.by.formality)){
        data.table::setorder(out, -F)
    }

    class(out) <- c("Formality", class(out))
    attributes(out)[["group.vars"]] <- G
    attributes(out)[["pos.vars"]] <- cols
    attributes(out)[["counts"]] <- formality_counts
    out

}


#' @export
#' @method formality Formality
formality.Formality <- function(text.var, grouping.var = NULL, order.by.formality = TRUE, ...){

    .SD <- n <- F <- NULL

    if(is.null(grouping.var)) {
        G <- "all"
    } else {
        if (is.list(grouping.var)) {
            m <- unlist(as.character(substitute(grouping.var))[-1])
            G <- sapply(strsplit(m, "$", fixed=TRUE), function(x) {
                    x[length(x)]
                }
            )
        } else {
            if (isTRUE(grouping.var)) {
                G <- "id"
            } else {
                G <- as.character(substitute(grouping.var))
                G <- G[length(G)]
            }
        }
    }
    if(is.null(grouping.var)){
        grouping <- rep("all", length(text.var))
    } else {
        if (isTRUE(grouping.var)) {
                grouping <- seq_along(text.var)
        } else {
            if (is.list(grouping.var) & length(grouping.var)>1) {
                grouping <- grouping.var
            } else {
                grouping <- unlist(grouping.var)
            }
        }
    }

    formal <- c('noun', 'preposition', 'adjective', 'article')
    contextual <- c('verb', 'pronoun', 'adverb', 'interjection')

    counts <- attributes(text.var)[["counts"]][["counts"]]

    counts <- cbind(
        data.table::as.data.table(stats::setNames(data.frame(grouping, stringsAsFactors = FALSE), G)),
        counts
    )

    out <- counts[, lapply(.SD, sum, na.rm = TRUE), keyby = G][,
        n := formal + contextual][,
        F := 50*(((formal - contextual)/(n))+1)]

    if (isTRUE(order.by.formality)){
        data.table::setorder(out, -F)
    }

    class(out) <- c("Formality", class(out))
    attributes(out)[["group.vars"]] <- G
    attributes(out)[["pos.vars"]] <- attributes(text.var)[["pos.vars"]]
    attributes(out)[["counts"]] <- attributes(text.var)[["counts"]][["counts"]]
    out

}



#' Plots a Formality Object
#'
#' Plots a Formality object.
#'
#' @param x The Formality object
#' @param plot logical.  If \code{TRUE} the output is plotted.
#' @param \ldots ignored.
#' @return Returns a list of the three \pkg{ggplot2} objects that make the
#' combined plot.
#' @importFrom data.table :=
#' @method plot Formality
#' @export
plot.Formality <- function(x, plot = TRUE, ...){

    group.vars <- n <- warn <- contextual <- formal <- type <- NULL

    grps <- attr(x, "group.var")
    pos <- attr(x, "pos.vars")

    ## Prepare the pos data
    express1 <- paste0("lapply(list(", paste(pos, collapse=","), "), function(y) as.numeric(y/n))")
    express2 <- paste0("paste(", paste(grps, collapse=", "), ", sep = \"_\")")
    pos_dat <- x[, c(grps, pos, "n"), with=FALSE][,
        (pos) := eval(parse(text=express1))][,
        'group.vars' := eval(parse(text=express2))][,
        'group.vars' := factor(group.vars, levels=rev(group.vars))][,
        c(pos, "n", "group.vars"), with = FALSE]

    pos_dat_long <- data.table::melt(pos_dat, id = c("group.vars", "n"),
        variable.name = "pos", value.name = "proportion")[,
        pos := factor(pos, levels = attr(x, "pos.vars"))]

    ## prepare the formality data
    form_dat <- x[, c(grps, "n", "F"), with=FALSE][,
        'group.vars' := eval(parse(text=express2))][,
        'group.vars' := factor(group.vars, levels=rev(group.vars))][,
        c("group.vars", "n", "F"), with = FALSE][,
        warn := factor(ifelse(n > 300, FALSE, TRUE), levels = c(TRUE, FALSE))]

    ## prepare the contectual/formal data
    con_form_dat <- x[, c(grps, "contextual", "formal", "n"), with=FALSE][,
        (c("contextual", "formal")) := list(contextual/n, formal/n)][,
        'group.vars' := eval(parse(text=express2))][,
        'group.vars' := factor(group.vars, levels=rev(group.vars))][,
        c("contextual", "formal", "n", "group.vars"), with = FALSE]

    con_form_long <- data.table::melt(con_form_dat, id = c("group.vars", "n"),
        variable.name = "type", value.name = "proportion")[,
        type := factor(type, levels = c("formal", "contextual"))]

    con_form_plot <- ggplot2::ggplot(con_form_long,
        ggplot2::aes_string(x = "group.vars", weight = "proportion", fill ="type")) +
        ggplot2::geom_bar() +
        ggplot2::coord_flip() +
        ggplot2::xlab(NULL) +
        ggplot2::ylab("") +
        ggplot2::theme_bw() +
        ggplot2::theme(
            panel.grid = ggplot2::element_blank(),
            #legend.position="bottom",
            legend.title = ggplot2::element_blank(),
            panel.border = ggplot2::element_blank(),
            axis.line = ggplot2::element_line(color="grey70")
        ) +
        ggplot2::scale_y_continuous(labels=function(x) paste0(round(x*100, 0), "%"),
           expand = c(0,0)) +
        ggplot2::scale_fill_manual(values=pals[c(2, 6), 2])

    form_plot <- ggplot2::ggplot(form_dat,
        ggplot2::aes_string(y = "group.vars", x = "F")) +
        ggplot2::geom_point(ggplot2::aes_string(size="n"), alpha=.22) +
        ggplot2::scale_size(range=c(1, 7), name = "Text\nLength") +
        ggplot2::geom_point(ggplot2::aes_string(color="warn"), size=1.5) +
        ggplot2::scale_color_manual(values=c("red", "black"), guide=FALSE, drop=FALSE) +
        ggplot2::ylab(NULL) +
        ggplot2::xlab("F Measure") +
        ggplot2::theme_bw() +
        ggplot2::theme(
            #legend.position="bottom",
            axis.title.x = ggplot2::element_text(size=11),
            #legend.title = ggplot2::element_blank(),
            panel.border = ggplot2::element_blank(),
            axis.line = ggplot2::element_line(color="grey70")
        )

     pos_heat_plot <- ggplot2::ggplot(pos_dat_long,
        ggplot2::aes_string(y = "group.vars", x = "pos", fill="proportion")) +
        ggplot2::geom_tile() +
        ggplot2::scale_fill_gradient(
            labels=function(x) paste0(round(x*100, 0), "%"),
            high="#90A19E",
            low="white",
            name = ggplot2::element_blank()
        )+
        ggplot2::ylab(NULL) +
        ggplot2::xlab("Part of Speech") +
        ggplot2::theme_bw() +
        ggplot2::theme(
            panel.grid = ggplot2::element_blank(),
            #legend.position="bottom",
            axis.title.x = ggplot2::element_text(size=11),
            legend.title = ggplot2::element_blank(),
            panel.border = ggplot2::element_rect(color="grey88"),
            axis.text.x = ggplot2::element_text(colour=names(col_key)[match(col_key, pos)])
        ) +
        ggplot2::guides(fill = ggplot2::guide_colorbar(barwidth = .5, barheight = 10)) #+
        #ggplot2::guides(fill = ggplot2::guide_colorbar(barwidth = 14, barheight = .5))


    plotout1 <- gridExtra::arrangeGrob(con_form_plot, form_plot,
        widths = grid::unit(c(.5, .5), "native"), ncol=2)

    plotout2 <- gridExtra::arrangeGrob(plotout1, pos_heat_plot, ncol=1)
    if (isTRUE(plot)) gridExtra::grid.arrange(plotout2)
    return(invisible(list(formality = form_plot, contextual_formal = con_form_plot, pos = pos_heat_plot)))
}


pals <- structure(list(pos = c("noun", "adjective", "preposition", "article",
    "pronoun", "verb", "adverb", "interjection"), cols = c("#8C510A",
    "#BF812D", "#DFC27D", "#F6E8C3", "#C7EAE5", "#80CDC1", "#35978F",
    "#01665E")), .Names = c("pos", "cols"), row.names = c(NA, -8L
    ), class = "data.frame")


col_key <- setNames(pals[["pos"]], c(rep('#BF812D', 4), rep('#01665E', 4)))


